include/nbl/builtin/glsl/virtual_texturing/impl_functions.glsl

```diff
@@ -118,13 +119,15 @@ vec4 nbl_glsl_vTextureGrad_impl(in uint formatID, in vec3 virtualUV, in mat2 dOr
 	// are we performing minification
 	bool positiveLoD = LoD>0.0;
 	// magnification samples LoD 0, else clip to max representable in VT
-	int clippedLoD = positiveLoD ? min(LoD_high,originalMaxFullMip):0;
+	int clippedLoD = positiveLoD ? min(LoD_high,originalMaxFullMip):0; // originalMaxFullMip is always -1 in case of no miplevel taking at least 1 full page
```

Comment:
Got a better idea (assuming `originalMaxFullMip` can now be -1)
```glsl
// WARNING: LoD_high will round up when LoD negative, its not a floor
const int LoD_high = int(LoD);
	
bool haveToDoTrilinear = false;
int levelInTail = 0;
int clippedLoD = 0;
if (LoD>0.f) // minification
{
   const bool smallerThanPageSz = LoD_high>originalMaxFullMip;
   clippedLoD = smallerThanPageSz ? originalMaxFullMip:LoD_high;
   if (smallerThanPageSz)
      levelInTail = LoD_high-originalMaxFullMip;
   if (levelInTail<int(_NBL_VT_IMPL_PAGE_SZ_LOG2))
      haveToDoTrilinear = true;
   else
      levelInTail = int(_NBL_VT_IMPL_PAGE_SZ_LOG2);
}

// get the higher resolution mip-map level
vec3 hiPhysCoord = nbl_glsl_vTexture_helper(formatID,virtualUV,clippedLoD,levelInTail);
// get lower if needed (speculative execution, had to move divergent indexing to a single place)
vec3 loPhysCoord;
// speculative if (haveToDoTrilinear)
{
   // now we have absolute guarantees that both LoD_high and LoD_low are in the valid original mip range
	if (LoD_high!=originalMaxFullMip) // high mip tap not last before the tail
	   clippedLoD++;
	else
	   levelInTail++;
        levelInTail = min(levelInTail, int(_NBL_VT_IMPL_PAGE_SZ_LOG2));
	loPhysCoord = nbl_glsl_vTexture_helper(formatID,virtualUV,clippedLoD,levelInTail);
}
```

Reply:
doesnt work, tested

Reply:
make me a branch I'll test

Reply:
ok done by me